File: fee/definition/src/action.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use warp_erc1155::state::Balance;

use crate::error::ContractError;
use crate::state::{Fees, State};

#[derive(JsonSchema, Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Transfer {
    // pub from: Option<String>,
    pub to: String,
    pub nft_id: String,
    pub price: Balance,
}

#[derive(JsonSchema, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateFee {
    pub nft_base_id: String,
    pub fees: Fees,
    pub rate: u32,
}

#[derive(JsonSchema, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Configure {
    pub super_operator: Option<String>,
    pub operators: Option<Vec<String>>,
}

#[derive(JsonSchema, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Evolve {
    pub value: String,
}

#[derive(JsonSchema, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Batch {
    pub actions: Vec<Action>,
}

#[derive(JsonSchema, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum NftScarcity {
    Unique,
    Legendary,
    Epic,
    Rare,
}

#[derive(JsonSchema, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MintNft {
    pub scarcity: NftScarcity,
    pub ticker: Option<String>,
    pub fees: Fees,
    pub rate: u32,
}

#[derive(JsonSchema, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", tag = "function")]
pub enum Action {
    CreateFee(CreateFee),

    Transfer(Transfer),

    Configure(Configure),

    Evolve(Evolve),

    MintNft(MintNft),

    Batch(Batch),
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase", untagged)]
pub enum ReadResponse {
    Batch(Vec<ReadResponse>),
}

#[derive(Serialize, Deserialize)]
#[serde(untagged)]
pub enum HandlerResult {
    Write(State),
    Read(State, ReadResponse),
}

pub type ActionResult = Result<HandlerResult, ContractError>;

File: scarcity/definition/bindings/ts/Action.ts
export type Action =
  | {
      baseId: string;
      function: "editAttachedRoyalties";
      rate: number;
      royalties: {
        [k: string]: number;
      };
    }
  | {
      baseId: string;
      function: "attachRoyalties";
      rate: number;
      royalties: {
        [k: string]: number;
      };
    }
  | {
      baseId?: string | null;
      function: "mintNft";
      rate: number;
      royalties: {
        [k: string]: number;
      };
      scarcity: Scarcity;
    }
  | {
      from: string;
      function: "transfer";
      price: string;
      to: string;
      tokenId: string;
    }
  | {
      canEvolve?: boolean | null;
      custodian?: string | null;
      erc1155?: string | null;
      function: "configure";
      operators?: string[] | null;
      paused?: boolean | null;
      superOperators?: string[] | null;
    }
  | {
      function: "evolve";
      value: string;
    }
  | {
      actions: Action[];
      function: "batch";
    };
export type Scarcity = "unique" | "legendary" | "epic" | "rare";

/**
 * This type allows to restrict the type of an interaction to a specific action.
 *
 * Example:
 * ```typescript
 * const specificAction: Actions["specificAction"] = { function: "specificAction", foo: "bar" };
 * ```
 */
export type Actions = {
    [K in Action["function"]]: Action & { function: K };
};